### PR TITLE
Run parse conformance for all packages in Hub

### PR DESCRIPTION
**Pull request by chayac**
## Description

New scripts used to run parse conformance on all packages in Package Hub. These are all run manually/locally for now.

## Type of change

*   [ ] Bug fix (non-breaking change which fixes an issue)
*   [x] New feature (non-breaking change which adds functionality)
*   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
*   [ ] Documentation update

## Checklist

*   [x] Ran `uv tool ruff format --config pyproject.toml` 
*   [ ] If this is a bug fix:
    *   [ ] Updated integration tests with bug repro
    *   [ ] Linked to bug report ticket
*   [x] Added unit tests if needed
*   [x] Updated unit tests if needed
*   [x] Tests passed when run locally

### CODE REVIEW

File: packages/dbt_fusion_package_tools/src/dbt_fusion_package_tools/compatibility.py [hunk 1]
```diff
@@ -1,12 +1,43 @@
...
-class FusionCompatibility(DataClassDictMixin):
+class FusionCompatibility(DataClassJSONMixin):
     require_dbt_version_defined: Optional[bool] = None
     require_dbt_version_compatible: Optional[bool] = None
     fusion_parse: Optional[bool] = None
     dbt_verified: Optional[bool] = None
+
+
+@dataclass
+class FusionLogMessage(DataClassJSONMixin):
```
Comment (davidharting) [inline packages/dbt_fusion_package_tools/src/dbt_fusion_package_tools/compatibility.py]: Are all of these logs derived from protos? 

Should we consider a follow-up mini project to use generated types for all of this sort of serialization / deserialization of artifacts and logs?

File: tests/unit_tests/package_upgrades/test_check_parse_conformance.py [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+            package_version_download_url="https://codeload.github.com/test/tar.gz/v1.0.0",
+            latest_package_version_download_url=None,
+        )
+
+        assert result is not None
+        assert result.download_failed is True
+
+    @patch("dbt_fusion_package_tools.check_parse_conformance.run_conformance_for_version")
+    @patch("dbt_fusion_package_tools.check_parse_conformance.requests.get")
+    def test_fallback_to_latest_url(self, mock_get, mock_conformance):
```
Comment (davidharting) [inline tests/unit_tests/package_upgrades/test_check_parse_conformance.py]: I'm confused about this behavior. We are trying to establish conformance of a particular version of a particular package, right. Falling back to the latest version instead is going to have test conformance of a different package entirely, right? 

Or is this a going to test a fallback URL for the same package?

File: packages/dbt_fusion_package_tools/src/dbt_fusion_package_tools/scripts/conformance_output.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+import csv
+import json
+from collections import Counter, defaultdict
+from datetime import datetime
+from pathlib import Path
+from typing import Any
+
+OUTPUT_PATH: Path = Path.cwd() / "src" / "dbt_fusion_package_tools" / "scripts" / "output"
+CONFORMANCE_OUTPUT_FILE_NAME: str = "v101_conformance_output_20260128.json"
```
Comment (davidharting) [inline packages/dbt_fusion_package_tools/src/dbt_fusion_package_tools/scripts/conformance_output.py]: Instead of hardcoding this, could you construct it from FUSION_VERSION and run_date?

File: packages/dbt_fusion_package_tools/src/dbt_fusion_package_tools/scripts/conformance_output.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+Two CSVs are written to OUTPUT_PATH:
+    "package_conformance_summary_v{fusion_version}_{run_date}.csv"
+    "package_conformance_all_errors_v{fusion_version}_{run_date}.csv"
+package_conformance_summary: one line per package version
+package_conformance_all_errors: one line per error per package version
+"""
+
+import csv
+import json
+from collections import Counter, defaultdict
```
Comment (davidharting) [inline packages/dbt_fusion_package_tools/src/dbt_fusion_package_tools/scripts/conformance_output.py]: TIL about the Counter class!

File: packages/dbt_fusion_package_tools/src/dbt_fusion_package_tools/check_parse_conformance.py [hunk 5]
```diff
@@ -1,88 +1,149 @@
...
-from dbt_fusion_package_tools.dbt_package_version import DbtPackageVersion
-from dbt_fusion_package_tools.version_utils import VersionSpecifier
-from dbtlabs.proto.public.v1.events.fusion.invocation.invocation_pb2 import Invocation
-from dbtlabs.proto.public.v1.events.fusion.log.log_pb2 import LogMessage
-from google.protobuf import json_format
 
 console = Console()
 error_console = Console(stderr=True)
 
+_ERROR_PATH_REGEX = re.compile(r"(?:\.\./)*")
```
Comment (davidharting) [inline packages/dbt_fusion_package_tools/src/dbt_fusion_package_tools/check_parse_conformance.py]: I needed to ask Claude about this regex (i need to ask claude about almost all regexes 😅)

Consider adding a pydoc for this constant:

```
"""Matches leading "../" sequences from file paths,
Useful for normalizing normalizing relative paths in stack traces to their base form.
"""
```

(not sure if I have the intent of it correct yet - still reading the file)

File: packages/dbt_fusion_package_tools/src/dbt_fusion_package_tools/check_parse_conformance.py [hunk 6]
```diff
@@ -1,88 +1,149 @@
...
+        # if we do have a file, extract the archive
+        try:
+            extract_dir = Path(tmpdir) / "extracted"
+            extract_dir.mkdir(parents=True, exist_ok=True)
+
+            with tarfile.open(tar_path, "r:gz") as tar:
+                for entry in tar:
+                    if os.path.isabs(entry.name) or ".." in entry.name:
+                        raise ValueError("Illegal tar archive entry")
+                    tar.extract(entry, extract_dir)
```
Comment (davidharting) [inline packages/dbt_fusion_package_tools/src/dbt_fusion_package_tools/check_parse_conformance.py]: I think we should make a safe extract tarball utility to better protect against malicious packages. I know Hub is controlled but it is still arbitrary code download from the internet. 

Some ideas from claude for this:

---

⏺ - Path.resolve() canonicalizes the path, resolving .., ., and any existing symlinks to get the true
  destination
  - is_relative_to() is safer than string startswith() comparison, which can have edge cases like /tmp/foo
  matching /tmp/foobar (requires Python 3.9+)
  - Symlinks are blocked because an attacker can include a symlink pointing outside the extraction directory, then write a file through that symlink
  - Hardlinks are blocked for similar reasons - they can reference files outside the extraction directory
  - Only regular files and directories are allowed, rejecting device files, FIFOs, and other special types
  that could cause issues
  - For Python 3.12+, this can be simplified to tar.extractall(path, filter="data") which provides these
  protections built-in


```python
  def safe_extract_tarball(tar_path: str | Path, extract_dir: str | Path) -> None:
      """Safely extract a gzipped tarball to a directory with path traversal protection.

      This function extracts tar archive contents while guarding against common
      archive-based attacks including:
      - Path traversal via absolute paths or ".." components
      - Symlink attacks (symlink pointing outside extraction directory)
      - Hardlink attacks (hardlink referencing files outside extraction directory)
      - Special file types (devices, FIFOs) that could cause issues

      Args:
          tar_path: Path to the .tar.gz file to extract.
          extract_dir: Directory to extract contents into. Will be created if it
              doesn't exist.

      Raises:
          ValueError: If the archive contains entries that would extract outside
              the target directory, or contains disallowed entry types (symlinks,
              hardlinks, devices, etc.).
          tarfile.TarError: If the archive is invalid or corrupted.
          FileNotFoundError: If tar_path doesn't exist.

      Note:
          For Python 3.12+, consider using the built-in data filter instead:
          ``tar.extractall(extract_dir, filter="data")``

      Example:
          >>> safe_extract_tarball("package.tar.gz", "/tmp/extracted")
      """
      extract_dir = Path(extract_dir).resolve()
      extract_dir.mkdir(parents=True, exist_ok=True)

      with tarfile.open(tar_path, "r:gz") as tar:
          for member in tar.getmembers():
              _validate_tar_member(member, extract_dir)
              tar.extract(member, extract_dir)


  def _validate_tar_member(member: tarfile.TarInfo, extract_dir: Path) -> None:
      """Validate that a tar member is safe to extract.

      Args:
          member: The tar archive member to validate.
          extract_dir: The resolved extraction directory path.

      Raises:
          ValueError: If the member fails any security check.
      """
      # Block symlinks - they can point outside extraction directory and allow
      # subsequent file writes to follow the link
      if member.issym():
          raise ValueError(
              f"Tar archive contains symbolic link which is not allowed: {member.name!r}"
          )

      # Block hardlinks - they can reference files outside extraction directory
      if member.islnk():
          raise ValueError(
              f"Tar archive contains hard link which is not allowed: {member.name!r}"
          )

      # Only allow regular files and directories
      if not (member.isfile() or member.isdir()):
          raise ValueError(
              f"Tar archive contains unsupported entry type: {member.name!r}"
          )

      # Resolve the target path and verify it's within extract_dir
      # Using resolve() handles "..", ".", and normalizes the path
      target_path = (extract_dir / member.name).resolve()

      # is_relative_to() is cleaner than startswith() string comparison
      # and correctly handles edge cases
      if not target_path.is_relative_to(extract_dir):
          raise ValueError(
              f"Tar archive entry would extract outside target directory: {member.name!r}"
          )
```

File: tests/unit_tests/package_upgrades/test_check_parse_conformance.py [hunk 7]
```diff
@@ -0,0 +1,483 @@
...
+        # Mock the tarball download
+        mock_response = Mock()
+        mock_response.iter_content = Mock(return_value=[b"mock_tar_data"])
+        mock_get.return_value = mock_response
+
+        # Create a real tar file for testing
+        with tempfile.TemporaryDirectory() as tmpdir:
+            # Create a mock tarball content
+            import io
+            import tarfile
```
[the remark this answers is not in the document]
Comment (chayac) [reply tests/unit_tests/package_upgrades/test_check_parse_conformance.py]: Blame Claude 🤷 but don't see any reason why it needs to be here

File: tests/unit_tests/package_upgrades/test_package_hub_fusion_compatibility.py [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+
+            result = read_json_from_local_hub_repo(str(base_dir.parent), file_count_limit=2)
+
+            # Should have read at most 2 files
+            total_files = sum(len(v) for v in result.values())
+            assert total_files <= 2
+
+    def test_read_nonexistent_path(self):
+        """Test reading from nonexistent path."""
+        with warnings.catch_warnings(record=True) as w:
```
[the remark this answers is not in the document]
Comment (chayac) [reply tests/unit_tests/package_upgrades/test_package_hub_fusion_compatibility.py]: Blame Claude 🤷 (positive)

File: packages/dbt_fusion_package_tools/src/dbt_fusion_package_tools/check_parse_conformance.py (hunk 5 above)
Comment (chayac) [reply packages/dbt_fusion_package_tools/src/dbt_fusion_package_tools/check_parse_conformance.py]: There is a note in line 260 when it's used but basically yeah, it removes the tempdir part of the path because otherwise the error messages look like this which is annoying as hell:
```"While parsing config: Ignored unexpected key `\"assertions\"`. YAML path: `columns[3].assertions`.\n  --> ../../../../../../private/var/folders/xf/r3bz6sxx0m57976q8nb2d1fh0000gp/T/tmp0bnpllg5/models/examples/basic_example/basic_example_d_site.yml:10:9"```

File: packages/dbt_fusion_package_tools/src/dbt_fusion_package_tools/scripts/update_package_hub_json.py [hunk 9]
```diff
@@ -0,0 +1,187 @@
+"""Script to add parse conformance data to Package Hub.
+
+Run as a CLI: `uv run update-package-hub`
+
+Reads the `conformance_output.json` output from `package_hub_fusion_compatibility`.
+Adds or updates the Fusion compatibility info for package versions in a local Hub repo.
```
[the remark this answers is not in the document]
Comment (chayac) [reply packages/dbt_fusion_package_tools/src/dbt_fusion_package_tools/scripts/update_package_hub_json.py]: Currently, yeah. In the future this will run within Hubcap when it runs and then cuts its own PRs to Hub.

File: packages/dbt_fusion_package_tools/src/dbt_fusion_package_tools/scripts/update_package_hub_json.py [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+        error_console.log(f"No package hub output found for {package_name} version {version}")
+        return {}
+
+
+def update_hub_json(
+    original_json: dict[str, Any],
+    conformance_output: FusionConformanceResult,
+    fusion_version: str,
+) -> dict[str, Any]:
+    # this is done to ensure that the ordering matches the original so the git diff is minimized
```
[the remark this answers is not in the document]
Comment (chayac) [reply packages/dbt_fusion_package_tools/src/dbt_fusion_package_tools/scripts/update_package_hub_json.py]: Exactly - these are already massive diffs so without maintaining insert order it was basically impossible to tell what had actually changed.
